File: data/derived_cot/rq3_prompting/results/translation/r1/py_to_cpp/code_output/CalendarUtil.cpp
#include <vector>
#include <string>
#include <algorithm>
#include <ctime>
#include <utility>

class DateTime {
public:
    int year, month, day, hour, minute, second;

    DateTime(int y, int mo, int d, int h = 0, int mi = 0, int s = 0)
        : year(y), month(mo), day(d), hour(h), minute(mi), second(s) {}

    DateTime() : year(0), month(0), day(0), hour(0), minute(0), second(0) {}

    struct Date {
        int year, month, day;
        bool operator==(const Date& other) const {
            return year == other.year && month == other.month && day == other.day;
        }
    };

    Date date() const {
        return {year, month, day};
    }

    bool operator<(const DateTime& other) const {
        if (year != other.year) return year < other.year;
        if (month != other.month) return month < other.month;
        if (day != other.day) return day < other.day;
        if (hour != other.hour) return hour < other.hour;
        if (minute != other.minute) return minute < other.minute;
        return second < other.second;
    }

    bool operator>(const DateTime& other) const { return other < *this; }
    bool operator<=(const DateTime& other) const { return !(*this > other); }
    bool operator>=(const DateTime& other) const { return !(*this < other); }

    bool operator==(const DateTime& other) const {
        return year == other.year && month == other.month && day == other.day &&
               hour == other.hour && minute == other.minute && second == other.second;
    }

    bool operator!=(const DateTime& other) const { return !(*this == other); }

    DateTime operator+(int minutes) const {
        DateTime res = *this;
        res.minute += minutes;

        // carry minutes -> hours
        while (res.minute >= 60) {
            res.minute -= 60;
            res.hour++;
        }

        // carry hours -> days (with simplified month/year handling)
        while (res.hour >= 24) {
            res.hour -= 24;
            res.day++;
            // simple month overflow (leap year handled for February)
            int daysInMonth[] = {31,28,31,30,31,30,31,31,30,31,30,31};
            if (res.year % 4 == 0 && (res.year % 100 != 0 || res.year % 400 == 0))
                daysInMonth[1] = 29;
            if (res.day > daysInMonth[res.month - 1]) {
                res.day = 1;
                res.month++;
                if (res.month > 12) {
                    res.month = 1;
                    res.year++;
                }
            }
        }
        return res;
    }

    static DateTime now() {
        std::time_t t = std::time(nullptr);
        std::tm* tm = std::localtime(&t);
        return DateTime(tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
                        tm->tm_hour, tm->tm_min, tm->tm_sec);
    }
};

struct Event {
    DateTime date;
    DateTime start_time;
    DateTime end_time;
    std::string description;

    bool operator==(const Event& other) const {
        return date == other.date &&
               start_time == other.start_time &&
               end_time == other.end_time &&
               description == other.description;
    }
};

class CalendarUtil {
private:
    std::vector<Event> events;

public:
    CalendarUtil() {}

    void add_event(const Event& event) {
        events.push_back(event);
    }

    void remove_event(const Event& event) {
        auto it = std::find(events.begin(), events.end(), event);
        if (it != events.end()) {
            events.erase(it);
        }
    }

    std::vector<Event> get_events(const DateTime& date) {
        std::vector<Event> result;
        for (const auto& ev : events) {
            if (ev.date.date() == date.date()) {
                result.push_back(ev);
            }
        }
        return result;
    }

    bool is_available(const DateTime& start_time, const DateTime& end_time) {
        for (const auto& ev : events) {
            if (start_time < ev.end_time && end_time > ev.start_time) {
                return false;
            }
        }
        return true;
    }

    std::vector<std::pair<DateTime, DateTime>> get_available_slots(const DateTime& date) {
        std::vector<std::pair<DateTime, DateTime>> slots;
        DateTime start_time(date.year, date.month, date.day, 0, 0);
        DateTime end_time(date.year, date.month, date.day, 23, 59);

        while (start_time < end_time) {
            DateTime slot_end = start_time + 60; // 60 minutes
            if (is_available(start_time, slot_end)) {
                slots.push_back(std::make_pair(start_time, slot_end));
            }
            start_time = start_time + 60;
        }
        return slots;
    }

    std::vector<Event> get_upcoming_events(int num_events) {
        DateTime now = DateTime::now();
        std::vector<Event> upcoming;
        for (const auto& ev : events) {
            if (ev.start_time >= now) {
                upcoming.push_back(ev);
                if (static_cast<int>(upcoming.size()) == num_events) {
                    break;
                }
            }
        }
        return upcoming;
    }
};

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/java_to_py/CalendarUtil.py
import datetime
from typing import List, Tuple, Optional

class Event:
    def __init__(self, date: datetime.datetime, start_time: datetime.datetime, end_time: datetime.datetime, description: str):
        self.date = date
        self.start_time = start_time
        self.end_time = end_time
        self.description = description

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Event):
            return False
        return (self.date == other.date and 
                self.start_time == other.start_time and 
                self.end_time == other.end_time and 
                self.description == other.description)

    def __hash__(self) -> int:
        return hash((self.date, self.start_time, self.end_time, self.description))

class CalendarUtil:
    def __init__(self):
        self.events: List[Event] = []

    def add_event(self, event: Event) -> None:
        self.events.append(event)

    def remove_event(self, event: Event) -> None:
        if event in self.events:
            self.events.remove(event)

    def get_events(self, date: datetime.datetime) -> List[Event]:
        target_date = date.date()
        return [event for event in self.events if event.date.date() == target_date]

    def is_available(self, start_time: datetime.datetime, end_time: datetime.datetime) -> bool:
        for event in self.events:
            if start_time < event.end_time and end_time > event.start_time:
                return False
        return True

    def get_available_slots(self, date: datetime.datetime) -> List[Tuple[datetime.datetime, datetime.datetime]]:
        available_slots = []
        start_time = date.replace(hour=0, minute=0, second=0, microsecond=0)
        end_time = date.replace(hour=23, minute=59, second=59, microsecond=999999)
        current = start_time
        while current < end_time:
            slot_end = current + datetime.timedelta(hours=1)
            if slot_end > end_time:
                slot_end = end_time
            if self.is_available(current, slot_end):
                available_slots.append((current, slot_end))
            current = slot_end
        return available_slots

    def get_upcoming_events(self, num_events: int) -> List[Event]:
        now = datetime.datetime.now()
        future_events = [event for event in self.events if event.start_time > now]
        future_events.sort(key=lambda event: event.start_time)
        return future_events[:num_events]